Signal/src/ViewControllers/OWSAddToContactViewController.m

```diff
@@ -0,0 +1,250 @@
+
+@implementation OWSAddToContactViewController
+
+- (instancetype)init
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    [self commonInit];
+
+    return self;
+}
+
+- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder
+{
+    self = [super initWithCoder:aDecoder];
+    if (!self) {
+        return self;
+    }
+
+    [self commonInit];
+
+    return self;
+}
+
+- (instancetype)initWithNibName:(nullable NSString *)nibNameOrNil bundle:(nullable NSBundle *)nibBundleOrNil
+{
+    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
+    if (!self) {
+        return self;
+    }
+
+    [self commonInit];
+
+    return self;
+}
+
+- (void)commonInit
+{
+    _contactsManager = [Environment getCurrent].contactsManager;
+    _contactsViewHelper = [[ContactsViewHelper alloc] initWithDelegate:self];
+}
+
+- (void)configureWithRecipientId:(NSString *)recipientId
+{
+    OWSAssert(recipientId.length > 0);
+
+    _recipientId = recipientId;
+}
+
+#pragma mark - ContactEditingDelegate
+
+- (void)didFinishEditingContact
+{
+    DDLogDebug(@"%@ %s", self.tag, __PRETTY_FUNCTION__);
+    [self dismissViewControllerAnimated:NO
+                             completion:^{
+                                 [self.navigationController popViewControllerAnimated:YES];
+                             }];
+}
+
+#pragma mark - CNContactViewControllerDelegate
+
+- (void)contactViewController:(CNContactViewController *)viewController
+       didCompleteWithContact:(nullable CNContact *)contact
+{
+    if (contact) {
+        // Saving normally returns you to the "Show Contact" view
+        // which we're not interested in, so we skip it here. There is
+        // an unfortunate blip of the "Show Contact" view on slower devices.
+        DDLogDebug(@"%@ completed editing contact.", self.tag);
+        [self dismissViewControllerAnimated:NO
+                                 completion:^{
+                                     [self.navigationController popViewControllerAnimated:YES];
+                                 }];
+    } else {
+        DDLogDebug(@"%@ canceled editing contact.", self.tag);
+        [self dismissViewControllerAnimated:YES
+                                 completion:^{
+                                     [self.navigationController popViewControllerAnimated:YES];
+                                 }];
+    }
+}
+
+#pragma mark - ContactsViewHelperDelegate
+
+- (void)contactsViewHelperDidUpdateContacts
+{
+    [self updateTableContents];
+}
+
+#pragma mark - View Lifecycle
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+
+    self.title = NSLocalizedString(@"CONVERSATION_SETTINGS_ADD_TO_EXISTING_CONTACT",
+        @"Label for 'new contact' button in conversation settings view.");
+
+    [self updateTableContents];
+}
+
+- (nullable NSString *)displayNameForContact:(Contact *)contact
+{
+    OWSAssert(contact);
+
+    if (contact.fullName.length > 0) {
+        return contact.fullName;
+    }
```

Comment:
We only let users add a phone number to a contact that already has name, phone number or email.